DOC: scale of parameters in optimize.curve_fit

PetteriAimonen opened the pull request:
#### Reference issue
Closes gh-19963

#### What does this implement/fix?
Added note about poor fit results if function parameters have large difference in scale.

PetteriAimonen committed:
DOC: scale of parameters in optimize.curve_fit (#19963)

Added note about poor fit results if function parameters
have large difference in scale.

j-bowhay commented:
Not at all a blocking comment as this is already an improvement, but if you can cook up a concise example this could be nice to demonstrate this in the examples section

PetteriAimonen committed:
DOC: add example of x_scale in optimize.curve_fit (#19963)

Shows how to use x_scale argument to handle functions with
large difference in parameter scales.

Renamed func() in redundant parameter example to func2() so that its
definition doesn't get mixed with the original func() in the first
example.

PetteriAimonen commented:
@j-bowhay Good idea, added now.

yagizolmez commented on scipy/optimize/_minpack_py.py:
```diff
@@ -864,6 +869,22 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
+    Fit errors can also be caused by scale differences between parameters.
+    An orders of magnitude difference between function parameters can give
+    inaccurate results or entirely fail to find a result.
+
```
If the optimal parameters of `f` differ by more than four orders of magnitude, the resulting fit can be inaccurate. Sometimes, `curve_fit()` can fail to find any results.

yagizolmez commented on scipy/optimize/_minpack_py.py:
```diff
@@ -790,6 +790,11 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
     Box constraints can be handled by methods 'trf' and 'dogbox'. Refer to
     the docstring of `least_squares` for more information.
 
+    Parameters to be fitted should have similar scale. Differences of multiple
```
According to the [documentation guidelines](https://docs.scipy.org/doc/scipy/dev/contributor/rendering_documentation.html#documentation-guidelines) you must use  `must` instead of `should` .

PetteriAimonen replied:
Thanks, I missed that.

yagizolmez commented on scipy/optimize/_minpack_py.py:
```diff
@@ -790,6 +790,11 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
     Box constraints can be handled by methods 'trf' and 'dogbox'. Refer to
     the docstring of `least_squares` for more information.
 
+    Parameters to be fitted should have similar scale. Differences of multiple
```
```suggestion
    Parameters to be fitted must have similar scale. Differences of multiple
```

PetteriAimonen committed:
Apply suggestions from code review

Co-authored-by: Yagiz Olmez <57116432+yagizolmez@users.noreply.github.com>

PetteriAimonen committed:
DOC: fix CI build error of curve_fit example

Have to catch the exception to avoid error in documentation build.
Also noticed that method = 'trf' tolerates larger scale errors, so
adjusted the example for that.

PetteriAimonen commented:
Is there anything further that I could do to help get this merged?

j-bowhay commented on scipy/optimize/_minpack_py.py:
```diff
@@ -864,6 +869,24 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
 
+    If the optimal parameters of `f` differ by more than four orders
+    of magnitude, the resulting fit can be inaccurate. 
+    Sometimes, `curve_fit()` can fail to find any results.
```
```suggestion
    Sometimes, `curve_fit` can fail to find any results.
```

j-bowhay commented on scipy/optimize/_minpack_py.py:
```diff
@@ -864,6 +869,24 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
+    Optimal parameters not found: The maximum number of function evaluations is exceeded.
+
+    If parameter scale is roughly known beforehand, it can be defined in
+    ``x_scale`` argument:
```
```suggestion
    `x_scale` argument:
```

j-bowhay reviewed (changes requested):
Sorry for the delay here, minor formatting nits then this should be good to go

PetteriAimonen committed:
Apply suggestions from code review

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

PetteriAimonen commented:
@j-bowhay Thanks, I agree with your suggestions and added those.

j-bowhay commented on scipy/optimize/_minpack_py.py:
```diff
@@ -864,6 +869,24 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
     suggesting that the optimal values of these parameters are ambiguous and
     that only one of these parameters is needed in the model.
 
+    If the optimal parameters of `f` differ by more than four orders
```
One final query, how did you arrive at four orders of magnitude here? Unless there is a concrete reason I'd perhaps prefer to stick to multiple orders of magnitude language used earlier

yagizolmez replied:
I believe this was my suggestion. I changed it from five to four based on the example. I agree that multiple orders of magnitude would be a better option.

PetteriAimonen replied:
Purely by experimentation. It seems to depend on a lot of things, and method=trf seems to be more robust against scale differences than the default method=lm. Generic "multiple orders" would be correct, though up to 4 orders of magnitude difference seems to work fine with any method.

j-bowhay replied:
I would prefer a more generic phrasing if that's ok as means we are not promising anything that we aren't 100% on

PetteriAimonen replied:
Done :)

PetteriAimonen committed:
Apply suggestions from review

j-bowhay reviewed (approved):
Thanks both and congratulations on your first contribution to SciPy, keep them coming!
